internal/pkg/crawler/imdbcrawler.go

```diff
@@ -0,0 +1,160 @@
+package imdb
+
+import (
+	"fmt"
+	"github.com/PhamDuyKhang/littledetective/internal/pkg/flog"
+	"github.com/PuerkitoBio/goquery"
+	"github.com/globalsign/mgo"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+type (
+	Film struct {
+		ID          string              `json:"id" bson:"_id"`
+		Rank        int                 `json:"rank" bson:"rank"`
+		URL         string              `json:"url" bson:"url"`
+		Title       string              `json:"title" bson:"title"`
+		Rate        string              `json:"rate" bson:"rate"`
+		ReleaseDate int                 `json:"release_date" bson:"release_date"`
+		Description string              `json:"description" bson:"description"`
+		Credit      map[string][]string `json:"credit" bson:"credit"`
+	}
+)
+
+func Crawler() {
+	l := flog.New()
+	l.SetLocal("imdb")
+	filmIn := make(chan Film, 20)
+	filmOut := make(chan Film, 20)
+	go func() {
+		var wg sync.WaitGroup
+		for i := 1; i <= 20; i++ {
+			wg.Add(1)
+			go ExtractDetail(i, wg, filmIn, filmOut, l)
+		}
+		wg.Wait()
+	}()
+	go func() {
+		SaveData(filmOut, l)
+	}()
+	l.Infof("extraction is stared")
+	MakeURLTopRate(filmIn, l)
+	l.Infof("extraction is done close channel")
+	close(filmIn)
+	return
+
+}
+func MakeURLTopRate(filmURLChan chan Film, l *flog.Flog) {
+	topURL := "https://www.imdb.com/chart/top?ref_=nv_mv_250"
+	doc, err := GetDocFormURL(topURL)
+	if err != nil {
+		fmt.Println(err)
+	}
+	doc.Find("table.chart").Find("tbody.lister-list").Find("tr").Each(func(rank int, films *goquery.Selection) {
+		f := Film{}
+		f.Rank = rank + 1
+		filmURL, ok := films.Find("td.titleColumn").Find("a").Attr("href")
+		if ok {
+			f.URL = NormalizeURL(filmURL)
+		}
+		f.Title = films.Find("td.titleColumn").Find("a").Text()
+		rawRelease := films.Find("td.titleColumn").Find("span.secondaryInfo").Text()
+		rawRelease = strings.TrimLeft(rawRelease, "(")
+		rawRelease = strings.TrimRight(rawRelease, ")")
+		releaseDate, err := strconv.Atoi(rawRelease)
+		if err != nil {
+			fmt.Println(err)
+			f.ReleaseDate = 0
+		} else {
+			f.ReleaseDate = releaseDate
+		}
+		f.Rate = films.Find("td.imdbRating").Find("strong").Text()
+		f.ID = NewUUID()
+		filmURLChan <- f
+	})
+
+}
+
+func ExtractDetail(idx int, wg sync.WaitGroup, fimIn chan Film, filmOut chan Film, l *flog.Flog) {
+	defer wg.Done()
+	for {
+		select {
+		case film, ok := <-fimIn:
+			if !ok {
+				l.Infof("crawler #%d done", idx)
+				return
+			}
+			l.Infof("crawler #%d is crawling", idx)
+			newFilm := film
```

Comment:
Why we have to copy the `film` to `newFilm` here?

Reply:
My mistake! When i write it, I'm not sure the `film` that I am received form channel. Which is value or reference, I worry my crawler is not correct, so I make a decisions, I copy new film, I want make `newFilm` is isolate form `film`